OpenRA.Mods.Common/Traits/RevealsShroud.cs

```diff
@@ -0,0 +1,103 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class RevealsShroudInfo : ITraitInfo
+	{
+		public readonly WRange Range = WRange.Zero;
+
+		public virtual object Create(ActorInitializer init) { return new RevealsShroud(init.Self, this); }
+	}
+
+	public class RevealsShroud : ITick, ISync, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly RevealsShroudInfo info;
+		readonly bool lobbyShroudFogDisabled;
+		[Sync] CPos cachedLocation;
+		[Sync] bool cachedDisabled;
+
+		protected Action<Player, CPos[]> add;
+		protected Action<Player> remove;
+		protected Func<bool> isDisabled;
+
+		public RevealsShroud(Actor self, RevealsShroudInfo info)
+		{
+			this.info = info;
+			lobbyShroudFogDisabled = !self.World.LobbyInfo.GlobalSettings.Shroud && !self.World.LobbyInfo.GlobalSettings.Fog;
+
+			add = (p, c) => p.Shroud.AddVisibility(self, c);
+			remove = p => p.Shroud.RemoveVisibility(self);
+			isDisabled = () => false;
+		}
+
+		static IEnumerable<CPos> FindVisibleTiles(World world, CPos position, WRange radius)
+		{
+			var map = world.Map;
+			var r = (radius.Range + 1023) / 1024;
+			var limit = radius.RangeSquared;
+			var pos = map.CenterOfCell(position);
+
+			foreach (var cell in map.FindTilesInCircle(position, r, true))
+				if ((map.CenterOfCell(cell) - pos).HorizontalLengthSquared <= limit)
+					yield return cell;
+		}
+
+		CPos[] Tiles(Actor self)
+		{
+			return Shroud.GetVisOrigins(self)
+				.SelectMany(o => FindVisibleTiles(self.World, o, Range))
+				.Distinct().ToArray();
+		}
+
+		public void Tick(Actor self)
+		{
+			if (lobbyShroudFogDisabled || !self.IsInWorld)
+				return;
+
+			var disabled = isDisabled();
+			if (cachedLocation == self.Location)
+				return;
+
+			var visible = Tiles(self);
```

Comment:
Let's go with `var tiles` here, again because of the dual purpose of the code (I would have asked for `visibleTiles` otherwise anyway ;) ).